tools/afl-fuzzer/main.c

```diff
@@ -0,0 +1,476 @@
+	exit(EXIT_FAILURE);
+}
+
+int fuzzer_send_msg(struct fuzz *fuzzer)
+{
+	struct timespec timeout;
+	struct timeval tp;
+	int ret;
+
+	ipc_dump(&fuzzer->msg);
+
+	/* send msg */
+	ret = fuzzer->platform->send_msg(fuzzer, &fuzzer->msg);
+	if (ret < 0) {
+		fprintf(stderr, "error: message tx failed\n");
+		return ret;
+	}
+
+	/* wait for ipc reply */
+	gettimeofday(&tp, NULL);
+	timeout.tv_sec  = tp.tv_sec;
+	timeout.tv_nsec = tp.tv_usec * 1000;
+	timeout.tv_nsec += 300000000; /* 300ms timeout */
+
+	/* first lock the boot wait mutex */
+	pthread_mutex_lock(&ipc_mutex);
+
+	/* reset condition for next IPC message */
+	ipc_reply_recd = 0;
+	ret = 0;
+
+	/* now wait for mutex unlock and ipc_reply_recd set by IPC reply */
+	while (!ipc_reply_recd && !ret)
+		ret = pthread_cond_timedwait(&ipc_cond, &ipc_mutex, &timeout);
+
+	if (ret == ETIMEDOUT && !ipc_reply_recd) {
+		ret = -EINVAL;
+		fprintf(stderr, "error: IPC timeout\n");
+		ipc_dump_err(&fuzzer->msg);
+		pthread_mutex_unlock(&ipc_mutex);
+		// exit(0);
+	}
+
+	pthread_mutex_unlock(&ipc_mutex);
+
+	/*
+	 * sleep for 5 ms before continuing sending the next message.
+	 * This helps with the condition signaling. Without this,
+	 * the condition seems to always satisfy and
+	 * the fuzzer never waits for a response from the DSP.
+	 */
+	usleep(50000);
+
+	return ret;
+}
+
+int main(int argc, char *argv[])
+{
+	struct fuzz fuzzer;
+	int ret;
+	char *ipc_msg_file = NULL;
+	char *dsp_run_file = NULL;
+	char *platform_name = "byt";
+	char opt;
+	FILE *fp = NULL, *fp1 = NULL;
+
+	while ((opt = getopt(argc, argv, "m:d:")) != -1) {
+		switch (opt) {
+		case 'm':
+			ipc_msg_file = optarg;
+			break;
+		case 'd':
+			dsp_run_file = optarg;
+			break;
+		default:
+			exit(EXIT_FAILURE);
+		}
+	}
+
+	if (!ipc_msg_file) {
+		fprintf(stderr, "error: Need to provide ipc-msg file using -m option\n");
+		exit(EXIT_FAILURE);
+	}
+
+	if (!dsp_run_file) {
+		// TODO: Re-write error statement to make it clear
+		fprintf(stderr, "error: Need to provide dsp initializer file using -d option\n");
+		exit(EXIT_FAILURE);
+	}
+
+	fp1 = fopen(dsp_run_file, "r");
+
+	if (fp1 == NULL) {
+		fprintf(stderr, "error: Opening dsp initializer file %s\n",
+			dsp_run_file);
+		exit(EXIT_FAILURE);
+	}
+
+	int ch = fgetc(fp1);
+	int to_boot = ch;
+	fclose(fp1);
+
+	if (ch == EOF) {
+		fp1 = fopen(dsp_run_file, "w");
+
+		if (!fp1) {
+			fprintf(stderr, "error: opening dsp initializer file %s\n",
+				dsp_run_file);
+			exit(EXIT_FAILURE);
+		}
+
+		ch = 'y';
+		if (fputc(ch, fp1) != ch) {
+			fprintf(stderr, "error: writing to dsp initializer file %s\n",
+				dsp_run_file);
+			exit(EXIT_FAILURE);
+		}
+		fclose(fp1);
+	}
+
+	// Only run fork() if ch is 'y' else don't.
+	if (ch == 'y' && !fork()) {
+
+		char *newargv[] = { "/home/sof/work/qemu/build/xtensa-softmmu/qemu-system-xtensa",
+				    "-cpu", "baytrail", "-M", "adsp_byt", "-nographic", "-kernel",
+				    "/home/sof/work/sof.git/build_byt_gcc/src/arch/xtensa/sof-byt.ri",
```

Comment:
is this path in your system or inside the the docker container? Also this appears to be hard coded to baytrail, can we add a TODO to make it multi platform?